crates/biome_css_parser/src/syntax/at_rule/keyframes.rs

```diff
@@ -0,0 +1,205 @@
+                .enable_recovery_on_line_break(),
+            expected_identifier,
+        )
+        .is_ok()
+    {
+        CSS_KEYFRAMES_AT_RULE
+    } else {
+        CSS_BOGUS_AT_RULE
+    };
+
+    if parse_keyframes_block(p)
+        .or_recover(
+            p,
+            &ParseRecovery::new(CSS_BOGUS_BLOCK, BODY_RECOVERY_SET).enable_recovery_on_line_break(),
+            expected_block,
+        )
+        .is_err()
+    {
+        return Present(m.complete(p, CSS_BOGUS_AT_RULE));
+    }
+
+    Present(m.complete(p, kind))
+}
+const KEYFRAMES_NAME_RECOVERY_SET: TokenSet<CssSyntaxKind> = token_set![T!['{']];
+
+#[inline]
+fn parse_keyframes_block(p: &mut CssParser) -> ParsedSyntax {
+    if !p.at(T!['{']) {
+        return Absent;
+    }
+
+    let m = p.start();
+    p.expect(T!['{']);
+    CssKeyframesItemList.parse_list(p);
+    p.expect(T!['}']);
+
+    Present(m.complete(p, CSS_KEYFRAMES_BLOCK))
+}
+
+const KEYFRAMES_ITEM_LIST_RECOVERY_SET: TokenSet<CssSyntaxKind> =
+    KEYFRAMES_ITEM_SELECTOR_SET.union(token_set!(T!['}']));
+
+struct CssKeyframesItemList;
+
+impl ParseNodeList for CssKeyframesItemList {
+    type Kind = CssSyntaxKind;
+    type Parser<'source> = CssParser<'source>;
+    const LIST_KIND: Self::Kind = CSS_KEYFRAMES_ITEM_LIST;
+
+    fn parse_element(&mut self, p: &mut Self::Parser<'_>) -> ParsedSyntax {
+        parse_keyframes_item(p)
+    }
+
+    fn is_at_list_end(&self, p: &mut Self::Parser<'_>) -> bool {
+        p.at(T!['}'])
+    }
+
+    fn recover(
+        &mut self,
+        p: &mut Self::Parser<'_>,
+        parsed_element: ParsedSyntax,
+    ) -> RecoveryResult {
+        parsed_element.or_recover(
+            p,
+            &ParseRecovery::new(CSS_BOGUS_KEYFRAMES_ITEM, KEYFRAMES_ITEM_LIST_RECOVERY_SET),
+            expected_keyframes_item,
+        )
+    }
+}
+
+#[inline]
+fn parse_keyframes_item(p: &mut CssParser) -> ParsedSyntax {
+    if !p.at_ts(KEYFRAMES_ITEM_SELECTOR_SET) {
+        return Absent;
+    }
+
+    let m = p.start();
+    CssKeyframesSelectorList.parse_list(p);
+
+    if parse_declaration_list_block(p)
+        .or_recover(
+            p,
+            &ParseRecovery::new(CSS_BOGUS_BLOCK, BODY_RECOVERY_SET).enable_recovery_on_line_break(),
+            expected_block,
+        )
+        .is_err()
+    {
+        return Present(m.complete(p, CSS_BOGUS_AT_RULE));
+    }
+
+    Present(m.complete(p, CSS_KEYFRAMES_ITEM))
+}
+
+const KEYFRAMES_ITEM_SELECTOR_LIST_RECOVERY_SET: TokenSet<CssSyntaxKind> =
+    KEYFRAMES_ITEM_SELECTOR_SET.union(token_set!(T!['{']));
+
+struct CssKeyframesSelectorList;
+
+impl ParseSeparatedList for CssKeyframesSelectorList {
+    type Kind = CssSyntaxKind;
+    type Parser<'source> = CssParser<'source>;
+    const LIST_KIND: Self::Kind = CSS_KEYFRAMES_SELECTOR_LIST;
+
+    fn parse_element(&mut self, p: &mut Self::Parser<'_>) -> ParsedSyntax {
+        //TODO better error recovery? we need to extend ParseRecovery to support nth_at token
```

Comment:
I've implemented basic error cases for keyframes. 
I guess we need to modify `ParseRecovery` because here we need to know not only the current token from the token set but also whether the next token is `%`.